src/connection.ts

```diff
@@ -0,0 +1,197 @@
+  func: (conn: Connection) => Promise<any>,
+  transActionConnection?: Connection
+): Promise<any> {
+  if (transActionConnection) {
+    return func(transActionConnection);
+  }
+  const conn = await (await getPool()).getConnection();
+  await conn.beginTransaction();
+  try {
+    const funcReturn = await func(new Connection(conn));
+    await conn.commit();
+    return funcReturn;
+  } catch (err) {
+    logger.error(err);
+    await conn.rollback();
+    throw err;
+  } finally {
+    conn.release();
+  }
+}
+
+export class Connection {
+  conn: Pool | PoolConnection;
+
+  constructor(conn: Pool | PoolConnection) {
+    this.conn = conn;
+  }
+
+  /**
+   * Use query when doing SELECT. The expected result will be an array of RowDataPacket
+   * @param query
+   */
+  public async query<T = RowDataPacket>(query: Query): Promise<T[]> {
+    const [statement, args = []] = query;
+    this.logSQL(statement, args);
+    const response = await this.exec(statement, args);
+    if (!Array.isArray(response)) {
+      throw Error(
+        'Return type error. query() should only be used for SELECT. For INSERT, UPDATE, DELETE and SET use execute()'
+      );
+    }
+    return response as T[];
+  }
+
+  public async queryMany<T = RowDataPacket>(queries: Query[]): Promise<T[][]> {
+    return await bluebird.map(queries, query => this.query<T>(query), { concurrency: 50 });
+  }
+
+  /**
+   * Use execute when doing INSERT, UPDATE or DELETE. The expected result will be a ResultSetHeader
+   * @param query
+   */
+  public async execute(query: Query): Promise<ResultSetHeader> {
+    const [statement, args = []] = query;
+    this.logSQL(statement, args);
+    const response = await this.exec(statement, args);
+    if (Array.isArray(response)) {
+      throw Error(
+        'Return type error. execute() should only be used for INSERT, UPDATE, DELETE and SET. For SELECT use query()'
+      );
+    }
+    return response as ResultSetHeader;
+  }
+
+  public async executeMany(queries: Query[]): Promise<ResultSetHeader[]> {
+    return await bluebird.map(queries, query => this.execute(query), { concurrency: 50 });
+  }
+
+  /**
+   *  Implementation of the mysql2 execute function conserving the return types.
+   *
+   *  Return definition
+   *
+   *  SELECT -> [result, columnDefinitions] result = Array
+   *
+   *  INSERT / UPDATE / DELETE -> [result] result = ResultSetHeader
+   *
+   * @param sql
+   * @param args
+   */
+  public async exec(
+    sql: string,
+    args: QueryArg[]
+  ): Promise<RowDataPacket[] | RowDataPacket[][] | OkPacket | OkPacket[] | ResultSetHeader> {
+    try {
+      const [response] = await this.conn.execute(sql, args);
+      return response;
+    } catch (err) {
+      const sqlFormatted = sqlFormatter.format('format' in this.conn ? this.conn.format(sql, args) : sql);
+      logger.error(`${err} - failing SQL: ${sqlFormatted}`);
+      throw Error(err);
+    }
+  }
+
+  private logSQL(sql: string, args: QueryArg[]) {
```

Comment:
I would just put this outside the object like so 

```
function logSQL(sql: string, args: QueryArg[]): string {
  const sqlFormatted = formatSQL(sql, args);
  logger.debug(sqlFormatted);
  return sqlFormatted;
}
```

and add this code to util in MySQL folder
```
export const formatSQL = (sql: string, args: QueryArg[] = []): string => sqlstring.format(sql, args);

```

Reply:
maybe out both in util